contracts/transfer/src/wasm/callable.rs

```diff
@@ -0,0 +1,356 @@
+use dusk_pki::PublicKey;
+use dusk_poseidon::cipher::PoseidonCipher;
+use phoenix_core::{Message, Note, NoteType};
+
+impl<S: Store> Transfer<S> {
+    fn call(&mut self, call: Call) -> bool {
+        match call {
+            Call::SendToContractTransparent {
+                address,
+                value,
+                value_commitment,
+                pk,
+                spend_proof,
+            } => self.send_to_contract_transparent(
+                address,
+                value,
+                value_commitment,
+                pk,
+                spend_proof,
+            ),
+
+            Call::WithdrawFromTransparent { address, note } => {
+                self.withdraw_from_transparent(address, note)
+            }
+
+            Call::SendToContractObfuscated {
+                address,
+                message,
+                r,
+                pk,
+                crossover_commitment,
+                crossover_pk,
+                spend_proof,
+            } => self.send_to_contract_obfuscated(
+                address,
+                message,
+                r,
+                pk,
+                crossover_commitment,
+                crossover_pk,
+                spend_proof,
+            ),
+
+            Call::WithdrawFromObfuscated {
+                address,
+                message,
+                r,
+                pk,
+                note,
+                input_value_commitment,
+                spend_proof,
+            } => self.withdraw_from_obfuscated(
+                address,
+                message,
+                r,
+                pk,
+                note,
+                input_value_commitment,
+                spend_proof,
+            ),
+
+            // Recursion not allowed
+            Call::Execute { .. } => false,
+
+            Call::None => true,
+        }
+    }
+
+    fn send_to_contract_transparent(
+        &mut self,
+        address: BlsScalar,
+        value: u64,
+        value_commitment: JubJubAffine,
+        pk: JubJubAffine,
+        spend_proof: Vec<u8>,
+    ) -> bool {
+        // Build proof public inputs
+        let scalars = 1 + BlsScalar::SIZE;
+        let points = (1 + JubJubAffine::SIZE) * 2;
+
+        let mut pi = Vec::with_capacity(scalars + points);
+        let label = "transfer-send-to-contract-transparent";
+        internal::extend_pi_jubjub_affine(&mut pi, &value_commitment);
+        internal::extend_pi_jubjub_affine(&mut pi, &pk);
+        internal::extend_pi_bls_scalar(&mut pi, &BlsScalar::from(value));
+
+        //  1. v < 2^64
+        //  2. B_a↦ = B_a↦ + v
+        if self.add_balance(address, value).is_err() {
+            return false;
+        }
+
+        //  3. if a.isPayable() ↦ true then continue
+        //  TODO
+
+        //  4. verify(C.c, v, π)
+        // TODO
+        let vk = keys::stct();
+        let (_, _, _, _) = (pi, label, spend_proof, vk);
+
+        //  5. C ← C(0,0,0)
+        //  TODO
```

Comment:
File a tech-debt for this issue or mention this in the issue created for the tech-debt generated by this PR merge.

Reply:
Outdated; solved with last commit